Add Java & Spring Boot performance docs. (#2715)
getsentry/sentry-docs: src/platforms/java/common/performance/index.mdx

---
title: Performance Monitoring
description: "Learn more about how to configure our Performance integrations to get the best experience out of it."
---

<Alert level="warning" title="Note">
Performance Monitoring is available for the Sentry Java SDK version ≥ 4.0.
</Alert>

> Sentry allows you to monitor the performance of your application, showing you how latency in one service may affect another service, and letting you pinpoint exactly which parts of a given operation may be responsible. To do this, it captures distributed traces consisting of transactions and spans, which measure individual services and individual operations within those services, respectively. You can learn more about this model in our [Distributed Tracing](/product/performance/distributed-tracing/) docs. Performance Monitoring is available for the Sentry Java SDK version ≥ 4.0.

Reviewer: We added the note above

```suggestion
Sentry allows you to monitor the performance of your application, showing you how latency in one service may affect another service, and letting you pinpoint exactly which parts of a given operation may be responsible. To do this, it captures distributed traces consisting of transactions and spans, which measure individual services and individual operations within those services, respectively. You can learn more about this model in our [Distributed Tracing](/product/performance/distributed-tracing/) docs.
```

## Enabling Tracing

To get started automatically sending traces you can:

- Set a uniform sample rate for all transactions, by setting the <PlatformIdentifier name="traces-sample-rate" /> option in your SDK config to a number between `0` and `1`. (For example, to send 20% of transactions, set <PlatformIdentifier name="traces-sample-rate" /> to `0.2`.)

> - Control the sample rate dynamically, based on the transaction itself and the context in which it's captured, by providing a function to the <PlatformIdentifier name="traces-sampler" /> config option.

Reviewer: I got confused if we need both. Below we point out only 1 is enough.
I feel we need to make it more clear:

```suggestion
- Control the sample rate dynamically, based on the transaction itself and the context in which it's captured, by providing a function to the <PlatformIdentifier name="traces-sampler" /> config option.

Or alternatively:
```

As you're getting tracing set up, we recommend setting <PlatformIdentifier name="traces-sample-rate" /> to `1`, so all created transactions are sent to Sentry. Once you're done with testing, though, you'll probably want to consider either **lowering your <PlatformIdentifier name="traces-sample-rate" /> value, or switching to <PlatformIdentifier name="traces-sampler" />**, which will allow you to set the sample rate individually for each transaction. Without sampling, automatically-captured transactions can add up quickly. (The Spring Boot integration, for example, will send a transaction for every request made to the server.) Sampling allows you to send representative data without overwhelming either your system or your Sentry transaction quota.

You can learn more about the <PlatformIdentifier name="traces-sample-rate" /> and <PlatformIdentifier name="traces-sampler" /> options in <PlatformLink to="/performance/sampling/">Sampling Transactions</PlatformLink>.

## Manual Instrumentation

To manually instrument certain regions of your code, you can create a transaction to capture them.

> The following example creates a transaction for a scope that contains an expensive operation (for example, `processItem`), and sends the result to Sentry:
> 
> ```java
> import io.sentry.Sentry;
> import io.sentry.SentryTransaction;
> import io.sentry.SpanStatus;
> 
> SentryTransaction transaction = Sentry.startTransaction("transaction name");
> transaction.setOperation("task");
> try {
>     processItem();
> } catch (Exception e) {

Reviewer: Maybe a sort of "unique" name and point out it's important the transaction name is relevant/helps you understand what this is about. Shows in the UI in a prevalent way.

Also `processItem` is used below on the span example.

```suggestion
The following example creates a transaction for a scope that contains an expensive operation (for example, `processOrderBatch`), and sends the result to Sentry:

```java
import io.sentry.Sentry;
import io.sentry.SentryTransaction;
import io.sentry.SpanStatus;

// A good name for the transaction is key, to help identify what this is about
SentryTransaction transaction = Sentry.startTransaction("processOrderBatch()");
transaction.setOperation("task");
try {
    procesOrderBatch();
} catch (Exception e) {
```

    transaction.setThrowable(e);
    transaction.setSpanStatus(SpanStatus.INTERNAL_ERROR)
} finally {
    transaction.finish();
}
```

### Adding More Spans to the Transaction

The next example contains the implementation of the hypothetical `processItem` function called from the code snippet in the previous section. Our SDK can determine if there is currently an open transaction and add all newly created spans as child operations to that transaction. Keep in mind that each individual span also needs to be manually finished; otherwise, spans will not show up in the transaction.

In cases where you want to attach Spans to an already ongoing Transaction you can use `Sentry#getSpan`. This method will return a `SentryTransaction` in case there is a running Transaction or a `Span` in case there is already a running Span, otherwise it returns `null`.

You can choose the value of `operation` and `description`.

```java
import io.sentry.ISpan;
import io.sentry.Span;
import io.sentry.SpanStatus;

void processItem(Item item):

>     ISpan span = Sentry.getSpan();

Reviewer: This means the user would need to have bound the transaction to the scope, according to what I got from @HazAT so far. And we haven't documented that.

Do we need to show that on the previous example? Or if we're binding it automatically, I wonder how this works on Android: Not possible to have more than 1 transaction at the same time in the app?

Author: Transactions are bound to the scope https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Scope.java#L24

Author: @marandaneto maybe you can jump in and tell if/how this will work on Android

Reviewer 2: tbh, I think it'd be a nice addition to be able to have multiple transactions at once for Mobile as well, but I don't see this as a huge blocker, at the end, it's a single user doing all the things (for most of the Apps), we can just document this limitation and tell people to go around with Spans if they need to measure different things (this is possible, right)?

In most of the cases, there will be a single transaction per screen anyway and if they want to create their own Hub for having multiple transactions, so they do.

The only concern I have right now is, a Hub is not ready for usage the way how it is, when creating a new Hub out of a fresh SentryOptions, a few things are not set like the Serializer, it's NoOp by default, it's only set on Sentry.init static class, what's about fixing this limitation, so at least we give a useful alternative for people?

Author: Regarding Hub initialization, since I don't know the background behind it I can't tell what are the side-effects, but perhaps @bruno-garcia can shine some light.

Reviewer 2: well it's pretty simple, we've discussed this offline recently actually.

do Hub(SentryOptions(dsn)).captureEvent(SentryEvent()), nothing is sent due to missing default options on SentryOptions, it should work ootb, eg a Serializer is not set (NoOp) by default.

>     Span innerSpan = span.startChild();
>     innerSpan.setOperation("task");
>     innerSpan.setDescription(item.getName());

Reviewer: Should we have an overload for `startChild`? This is consistently taking 3 lines of code for all operations.

```suggestion
    Span innerSpan = span.startChild(item.getName(), "task");
```

Author: Addressed in: https://github.com/getsentry/sentry-java/pull/1073

    try {
        // omitted code
    } catch (NotFoundException e) {
        innerSpan.setThrowable(e);
        innerSpan.setSpanStatus(SpanStatus.NOT_FOUND)
    } finally {
        innerSpan.finish();
    }
}
```

### Retrieving a Transaction

In cases where you want to attach Spans to an already ongoing Transaction you can use `Sentry#getSpan`. This method will return a `SentryTransaction` in case there is a running Transaction or a `Span` in case there is already a running Span, otherwise it returns `null`.

```java
import io.sentry.ISpan;
import io.sentry.Span;
import io.sentry.SpanStatus;

> ISpan span = Sentry.getSpan();

Reviewer: This won't really work with asynchronous code (this relies on TLS) so should we add a section to the docs about that?

What happens if I want to capture a span for an outgoing request that's happening on a background thread?

Author: Yes there is a single active transaction per `Hub`. I would imagine background thread should use a separate hub instance

Reviewer 2: so if I want to have a transaction per App. screen, also being able to create spans and its children for background op (like outgoing requests), etc... having a Single Hub which is the case on Android, I don't see this as a problem unless I am missing something, is this described scenario possible to achieve with a single Hub?

Author: The limitation is that there can be single active transaction per hub

Reviewer 2: the question still remains:

> is the described scenario above achievable with a single Hub?

the question is probably yes, but I'd like a confirmation.

Author: It depends. If the transaction in the parent thread is finished before the child thread's span is finished, child thread's span won't be sent together with transaction.

Reviewer 2: I see, that makes sense, but for now, the support for performance on Android is "manual", so documenting this is just fine:

"finish your transaction only after your spans have finished".

Author: Thanks, updated!
